## Question 3: Is the permutation divisible by 7? (1 mark)

Given an integer string, create all integer permutations of its digits. Determine if there is a permutation whose integer value is evenly divisible by 7, i.e., `(permutation value) mod 7=0`.

For example, the possible permutations of 789 are `p = {789, 798, 879, 897, 978, 987}`. Of
these values, `p[2]` and `p[5]` is divisible by 7 because `879 mod 7 = 0` and `987 mod 7 = 0`.

Their average is `(879+897)/2=933`.

What you'll need to do is determine if any of the permutations of 1867 are divisible by 7, and if so, what is the average between the smallest and largest permutation? Decimals are allowed.

In [1]:
# find the possible permutations of the string
def permutations(N, i=0, p=None):
    if p is None:
        p = []
    if i == len(N):
        p.append(int("".join(N)))
    for j in range(i, len(N)):
        num = [d for d in N]
        num[i], num[j] = num[j], num[i]
        permutations(num, i+1, p)
    return p

p_list = permutations(str(1867))

# determine which permutations are divisible by 7
p_div = []
for k in range(len(p_list)):
    if p_list[k] % 7 == 0:
        p_div.append(p_list[k])

print("The following permuations are divisible by 7:", p_div)
        
# average the smallest and largest permutation among those divisible by 7
print(int(min(p_div) + max(p_div))/2)

The following permuations are divisible by 7: [1876, 1687, 8176, 8617, 7861, 7168]
5152.0


The average between the smallest and largest permutation is **5152**.